network/p2p/gossip/bloom.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"hash"
+
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+var _ hash.Hash64 = (*hasher)(nil)
+
+// NewBloomFilter returns a new instance of a bloom filter with at most
+// [maxExpectedElements] elements anticipated at any moment, and a false
+// positive probability of [falsePositiveProbability].
+func NewBloomFilter(
+	maxExpectedElements uint64,
+	falsePositiveProbability float64,
+) (*BloomFilter, error) {
+	bloom, err := bloomfilter.NewOptimal(
+		maxExpectedElements,
+		falsePositiveProbability,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	salt, err := randomSalt()
+	return &BloomFilter{
+		Bloom: bloom,
+		Salt:  salt,
+	}, err
+}
+
+type BloomFilter struct {
+	Bloom *bloomfilter.Filter
+	// Salt is provided to eventually unblock collisions in Bloom. It's possible
+	// that conflicting Gossipable items collide in the bloom filter, so a salt
+	// is generated to eventually resolve collisions.
+	Salt ids.ID
+}
+
+func (b *BloomFilter) Add(gossipable Gossipable) {
+	h := gossipable.GetID()
+	salted := &hasher{
+		hash: h[:],
+		salt: b.Salt,
+	}
+	b.Bloom.Add(salted)
+}
+
+func (b *BloomFilter) Has(gossipable Gossipable) bool {
+	h := gossipable.GetID()
+	salted := &hasher{
+		hash: h[:],
+		salt: b.Salt,
+	}
+	return b.Bloom.Contains(salted)
+}
+
+// ResetBloomFilterIfNeeded resets a bloom filter if it breaches a target false
+// positive probability. Returns true if the bloom filter was reset.
+func ResetBloomFilterIfNeeded(
+	bloomFilter *BloomFilter,
+	falsePositiveProbability float64,
+) (bool, error) {
+	if bloomFilter.Bloom.FalsePosititveProbability() < falsePositiveProbability {
+		return false, nil
+	}
+
+	// it's not possible for this to error assuming that the original
+	// bloom filter's parameters were valid
+	bloomFilter.Bloom, _ = bloomfilter.New(bloomFilter.Bloom.M(), bloomFilter.Bloom.K())
+
+	salt, err := randomSalt()
+	bloomFilter.Salt = salt
+
+	return true, err
```

Comment:
Should we guarantee that the filter isn't modified if an error is returned?

Reply:
Yeah we shouldn't mess with the filter if we can't make a new salt.